libraries/adaptive-expressions/src/triggerTrees/clause.ts

```diff
@@ -0,0 +1,285 @@
+/**
+ * @module adaptive-expressions
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Expression } from '../expression';
+import { ExpressionType } from '../expressionType';
+import { PredicateComparer } from './optimizer';
+import { RelationshipType } from './relationshipType';
+
+/**
+ * A canonical normal form expression.
+ */
+export class Clause extends Expression {
+    private _ignored: Expression;
+
+    /**
+     * Initializes a new instance of the `Clause` class.
+     * @param clauseOrExpression A clause, expression or an array of expressions to initialize a `Clause`.
+     */
+    public constructor(clauseOrExpression?: Clause | Expression | Expression[]) {
+        super(ExpressionType.And, undefined);
+        if (clauseOrExpression) {
+            if (Array.isArray(clauseOrExpression)) {
+                const children: Expression[] = clauseOrExpression;
+                this.children = children;
+            } else if (clauseOrExpression instanceof Clause) {
+                const fromClause: Clause = clauseOrExpression;
+                this.children = [...fromClause.children];
+                for (const key in fromClause.anyBindings) {
+                    this.anyBindings[key] = fromClause.anyBindings[key];
+                }
+            } else if (clauseOrExpression instanceof Expression) {
+                const expression: Expression = clauseOrExpression;
+                this.children.push(expression);
+            }
+        }
+    }
+
+    /**
+     * Gets or sets the anyBinding dictionary.
+     */
+    public anyBindings: { [key: string]: string } = {};
+
+    /**
+     * Gets or sets whether the clause is subsumed.
+     */
+    public subsumed = false;
+
+    /**
+     * Gets a string that represents the current clause.
+     * @param builder An array of string to build the string of clause.
+     * @param indent An integer represents the number of spaces at the start of a line.
+     */
+    public toString(builder: string[] = [], indent = 0): string {
+        builder.push(' '.repeat(indent));
+        if (this.subsumed) {
+            builder.push('*');
+        }
+
+        builder.push('(');
+        let first = true;
+        for (const child of this.children) {
+            if (first) {
+                first = false;
+            } else {
+                builder.push(' && ');
+            }
+            builder.push(child.toString());
+        }
+
+        builder.push(')');
+        if (this._ignored) {
+            builder.push(' ignored(');
+            builder.push(this._ignored.toString());
+            builder.push(')');
+        }
+
+        for (const key in this.anyBindings) {
```

Comment:
Not sure if our `tsconfig.json` currently supports [Object.entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) but I tend to prefer this pattern:

```typescript
Object.entries(this.anyBindings).forEach(([key, value]) => builder.push(` ${ key }->${ value }`))
```

Many of these style suggestions (Array.prototype.forEach, Array.prototype.map, etc.) are not requirements, I just tend to prefer them. It's a little more straightforward and obvious than the subtlety of `for...of` vs. `for...in`.